ls1intum/Artemis
# `Iris` Enhancements of UX in the Chatbot Widget

### Checklist
#### General


#### Server
#### Client
#### Changes affecting Programming Exercises

### Motivation and Context



### Description
This PR addresses various UX problems.
1. Scroll to the bottom button was fixed. How it won't appear immediately and will let you scroll approximately through 1 message.
2. Empty message div disappears completely, instead of setting opacity to 0.
3. Pyris exceptions are now transmitted via Artemis to the client.
4. When disabling Iris during interactions with the chatbot, a corresponding error message will be shown.
5. Iris messages should be synchronized across multiple browser tabs.

### Steps for Testing
Prerequisites:
- 1 Instructor
- 1 Students
- 1 Programming Exercise

1. Log in to Artemis
2. Navigate to Course Administration Section
3. Enable Iris on all levels
4. Open a an exercise page on 2 tabs
5. Go into the exercise view and start chatting with the LLM
6. Ensure the messages appear on both tabs, without any duplications
7. Disable Iris on exercise level while leaving the chat widget open in another tab
8. Send something, ensure a correct error message appears
9. Reenable Iris
10. Send 5-6 messages, make sure scrolling gets enabled. Make sure the arrow appears (not immediately).
11. If you can mock Pyris responses(locally or on Pyris VM), do so, and check that correct error messages appear. Alternatively, navigate to global Iris settings and modify the template to an invalid one.


### Review Progress






#### Performance Review
#### Code Review
#### Manual Tests
#### Exam Mode Test

### Test Coverage







### Screenshots
